azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -195,87 +213,94 @@ def _fetch_goal_state(wire_client):
 
         wire_client.update_host_plugin(container_id, role_config_name)
 
-        return xml_text, xml_doc, incarnation
+        return incarnation, xml_text, xml_doc
 
-    def _fetch_vm_settings(self, force_update=False):
+    @staticmethod
+    def _fetch_vm_settings(wire_client):
         """
-        Issues an HTTP request (HostGAPlugin) for the vm settings and returns the response as an ExtensionsGoalStateFromVmSettings.
+        Issues an HTTP request (HostGAPlugin) for the vm settings and returns the response as an ExtensionsGoalState.
         """
         vm_settings, vm_settings_updated = (None, False)
 
         if conf.get_enable_fast_track():
             try:
-                vm_settings, vm_settings_updated = self._wire_client.get_host_plugin().fetch_vm_settings(force_update=force_update)
+                try:
+                    vm_settings, vm_settings_updated = wire_client.get_host_plugin().fetch_vm_settings()
+                except ResourceGoneError:
+                    # retry after refreshing the HostGAPlugin
+                    GoalState.update_host_plugin_headers(wire_client)
+                    vm_settings, vm_settings_updated = wire_client.get_host_plugin().fetch_vm_settings()
 
             except VmSettingsNotSupported:
                 pass
-            except VmSettingsError as exception:
-                # ensure we save the vmSettings if there were parsing errors
-                self._history.save_vm_settings(ExtensionsGoalStateFromVmSettings.redact(exception.vm_settings_text))
+            except VmSettingsParseError as exception:
+                # ensure we save the vmSettings if there were parsing errors, but save them only once per ETag
+                if not GoalStateHistory.tag_exists(exception.etag):
+                    GoalStateHistory(timeutil.create_timestamp(), exception.etag).save_vm_settings(exception.vm_settings_text)
```

Comment:
If by some magic the parsing error goes away in the next iteration for the same etag, will we overwrite the same history directory? Would we want to keep the bad file for debugging purposes?

Reply:
the timestamp would be different in the next iteration

that is the reason i am saving this only once per goal state, otherwise we would be creating different directories (since the timestamp is different) until we get a new goal state that clears the error

also, the directories for those bad settings end in "-etag", vs "_incarnation-etag" for the regular case

Reply:
Ahh makes sense. Thanks for clarifying